File: pwmdemo/pwmexplorer/apps/freqtest.go
//
// Copyright (c) 2021 Richard Youngkin. All rights reserved.
// Use of this source code is governed by a MIT-style
// license that can be found in the LICENSE file.
//
// This program visually demonstrates the the association between rpio.SetFreq() and
// rpio.SetDutyCycle(). rpio.SetFreq() sets the PWM clock frequency. rpio.SetDutyCycle()
// specifies how long a pin is in HIGH state (pulsewidth) vs. the length of the containing
// range.
//
// The Raspberry Pi 3B+ has a number of available clocks. The go-rpio library uses a clock called
// the Oscillator. This clock has a frequency of 19.2MHz. This freqency can be stepped down to a
// frequency which is more suitable for a given device like a motor which might require an input
// frequency of 100kHz. Using the go-rpio library it is possible to directly set the desired frequency
// of the PWM clock using 'SetFreq()'.
//
// Given a PWM clock frequency of 100,000 and a range length of 25,000, the LED frequency is 4Hz,
// that is it blinks 4 times per second. If the range length is changed to 400,000, the
// LED frequency is 0.25Hz, or once every 4 seconds. And finally, if the range length is 1,000
// the LED frequency is 100Hz. At 100Hz, the LED will appear to be continuously on, i.e., not
// blinking.  Using the go-rpio library range and pulse width are set indirectly by setting the
// duty cycle using 'SetDutyCycle()'.
//
// Run example (taking default values for PWM parameters):
// sudo /usr/local/go/bin/go run apps/freqtest.go
// Run example (See 'flag.*Var()' calls below for details regarding the meaning of the flags):
// sudo /usr/local/go/bin/go run apps/freqtest.go -pwmType="hardware" -freq="5000" -range="100" -pulsewidth="2" -pin="18"
//

package main

import (
	"flag"
	"fmt"
	"log"
	"os"
	"os/signal"
	"syscall"
	"time"

	"github.com/stianeikeland/go-rpio/v4"
)

func main() {
	var (
		freq       int
		rrange     int
		pwmPin     int
		pulsewidth int
		pwmMode    bool
	)
	pwmType := ""

	flag.StringVar(&pwmType, "pwmType", "hardware", "defines whether software or hardware PWM should be used")
	flag.IntVar(&freq, "freq", 5000, "desired PWM clock frequency")
	flag.IntVar(&rrange, "range", 10, "PWM range")
	flag.IntVar(&pwmPin, "pin", 18, "BCM pin number")
	flag.IntVar(&pulsewidth, "pulsewidth", 2, "PWM Pulse Width")
	flag.BoolVar(&pwmMode, "pwmmode", rpio.Balanced, "PWM Mode, Balanced (true) or MarkSpace (false)")
	flag.Parse()

	fmt.Printf("Using: PWM pin: %d, PWM Type: %s, freq: %d, range: %d, pulse width: %d, pwm mode: %t\n",
		pwmPin, pwmType, freq, rrange, pulsewidth, pwmMode)

	// Obtain GPIO resources
	if err := rpio.Open(); err != nil {
		log.Fatal(err)
		os.Exit(1)
	}
	defer rpio.Close()
	if pwmType == "software" {
		runSoftwarePWM(pwmPin, rrange, pulsewidth)
		return
	}

	runHardwarePwm(pwmPin, freq, uint32(rrange), uint32(pulsewidth), pwmMode)
}

// runHardwarePWM starts PWM as implemented in the board hardware.
func runHardwarePwm(gpin, freq int, rrange, pulsewidth uint32, pwmMode bool) {
	// Define the pin to be used,
	// followed by setting its mode to PWM,
	// then set directly set the PWM Clock frequency (note lack of divisor),
	// finally set the range and pulse width (pin.DutyCycle())
	// and send the PWM signal to the pin
	pin := rpio.Pin(gpin)
	pin.Mode(rpio.Pwm)
	pin.Freq(freq)
	// To test rpio.SetDutyCycle() comment out the next line and uncomment the one below it.
	rpio.SetDutyCycleWithPwmMode(pin, pulsewidth, rrange, pwmMode)
	//rpio.SetDutyCycle(pin, pulsewidth, rrange)

	// Initialize signal handling needed to catch ctl-C
	sigs := make(chan os.Signal, 1)
	signal.Notify(sigs, syscall.SIGINT, syscall.SIGKILL)
	go hardwareInterruptHandler(sigs, rrange, pin)

	// Sleep until ctl-C is caught
	for {
		time.Sleep(time.Millisecond * 20)
	}
}

// runSoftwarePWM starts PWM implemented by this function vs. via the board hardware
func runSoftwarePWM(gpin, rrange, pulsewidth int) {
	// Define the pin to be used for software PWM
	pin := rpio.Pin(gpin)
	// Set the pin mode to output so it can accept write
	// requests.
	pin.Output()

	// Initialize signal handling needed to catch ctl-C
	sigs := make(chan os.Signal, 1)
	signal.Notify(sigs, syscall.SIGINT, syscall.SIGKILL)
	go softwareInterruptHandler(sigs, rrange, pin, rpio.Low)

	for {
		// Execute the software driven duty cycle defined by rrange and
		// pulsewidth.
		//
		// The ratio of time the pin is set to 'on' vs. 'off' determines how bright the LED
		// will be. Lower pulsewidths will result in a dimmer LED. Lower pulsewidths (e.g., 10)
		// may also, depending on the chosen rage, exhibit more flickering vs. higher
		// values (e.g., 10000).
		//
		// The `time.Sleep` multipler is 100. Given the units are microseconds (time.Microsecond)
		// the value of 100 results in a period of 100 microseconds. This equates to a
		// frequency of 10kHz (1/0.0001). This frequency is fixed and can only be changed by
		// modifying this code.
		//
		// Notice that the LED might flicker due to a lack of uniformity in the OS scheduler
		// and the 'time.Sleep()' function. The amount of actual flicker that may occur will
		// depend on the computing power of the hardware and the CPU utilization.
		//
		// Loop until a ctl-C signal is caught
		for {
			rpio.WritePin(pin, rpio.High)
			time.Sleep((time.Microsecond * 100) * time.Duration(pulsewidth))
			rpio.WritePin(pin, rpio.Low)
			time.Sleep((time.Microsecond * 100) * time.Duration(rrange-pulsewidth))

		}
	}
}

// hardwareInterruptHandler handles the SIGINT caught when ctl-C is entered. It will
// return the pin to it's state before the program started and release GPIO resources
// before exiting. It is registered as the SIGINT handler when hardware PWM is
// specified.
func hardwareInterruptHandler(sigs chan os.Signal, rrange uint32, pin rpio.Pin) {
	<-sigs
	fmt.Println("\nExiting...")
	// Turn off the LED
	pin.DutyCycle(0, rrange)
	//pin.Mode(rpio.Output)
	rpio.Close()
	os.Exit(0)

}

// softwareInterruptHandler handles the SIGINT caught when ctl-C is entered. It will
// return the pin to it's state before the program started and release GPIO resources
// before exiting. It is registered as the SIGINT handler when softare PWM is specified.
func softwareInterruptHandler(sigs chan os.Signal, rrange int, pin rpio.Pin, off rpio.State) {
	<-sigs
	fmt.Println("\nExiting...")
	// Turn off the LED
	rpio.WritePin(pin, off)
	rpio.Close()
	os.Exit(0)

}

File: pwmdemo/pwmexplorer/apps/freqtest.c
//
// Copyright (c) 2021 Richard Youngkin. All rights reserved.
// Use of this source code is governed by a MIT-style
// license that can be found in the LICENSE file.
//
// This program visually demonstrates the association between the wiringPi functions
// 'pwmSetClock()', 'pwmSetRange()' and 'pwmWrite()'. 
//
// 'pwmSetClock' sets the divisor used to scale back the BCM2835 clock input to
// the desired PWM clock frequency. NOTE: The WiringPi library limits the parameter
// to 'pwmSetClock()', called the divisor, to 4095. It uses '&=' to accomplish this,
// which means a value of 4096 equates to a divisor of 0 and 4097 equates to a divisor
// 1, and so on. 'pwmSetRange()' sets the range for the PWM duty cycle. 'pwmWrite()' 
// sets the pulsewidth for the PWM duty cycle. Duty cycle is the ratio between the
// pulsewidth and the range. 
//
// Given a PWM clock frequency of 100,000 and a range length of 25,000, the LED frequency is 4Hz,
// that is it blinks 4 times per second. If the range length is changed to 400,000, the
// LED frequency is 0.25Hz, or once every 4 seconds. And finally, if the range length is 1,000
// the LED frequency is 100Hz. At 100Hz, the LED will appear to be continuously on, i.e., not
// blinking.
//
// NOTE: There is inadequate error handling in the application. Take care when copying.
// 
// Build: gcc -o freqtest freqtest.c  -lwiringPi -lpthread
// Run: sudo apps/freqtest --pin=<pwmPinNo> --divisor=<2 to 4095> --range=<n> --pulsewidth=<n> --type=<hardware|software> --mode=<balanced|markspace>
// E.g., sudo apps/freqtest --pin=1 --divisor=192 --range=1000 --pulsewidth=50 --type=hardware --mode=balanced
// E.g., or just sudo apps/freqtest if defaults are acceptable
// 
#include <wiringPi.h>
#include <softPwm.h>
#include <stdio.h>
#include <getopt.h>
#include <stdlib.h>
#include <signal.h>
#include <unistd.h>
#include <string.h>
 
// define magic values
int globalPin = 0;
char* markspace = "markspace";
char* balanced = "balanced";
char* hardware = "hardware";
char* software = "software";
int   OFF      = 0;
int   ON     = 1;

// Define interrupt handlers
void softwareInterruptHandler(int);
void hardwareInterruptHandler(int);

// Define functions to run hardware or software PWM
void runHardwarePWM(char* pwmMode, int divisor, int range, int pin, int pulsewidth); 
void runSoftwarePWM(int pin, int range, int pulsewidth); 

void printUsage() {
    printf("\nUsage:\n");
    printf("\t--help (prints this message)\n");
    printf("\t--type=[software|hardware]\n");
    printf("\t--mode=[markspace|balanced]\n");
    printf("\t--divisor=[2 to 4095]\n");
    printf("\t--range=[n]\n");
    printf("\t--pulsewidth=[0 to 'range']\n");
    printf("\t--pin=[7|21|22|26|23|27|0|1|24|28|29|3|4|5|6|25|2]\n");
    printf("\t\tHardware PWM pins are 0, 1, 23, and 24\n");
}


int main(int argc, char *argv[]) {
    // set default values for options  
    char *pwmType = hardware;
    char *pwmMode = balanced;
    int divisor = 192;
    int range = 1000;
    int pin = 1;
    int pulsewidth = 50;
                                                                     
    int c;
    int help_flag;

    static struct option long_options[] =
    {
        {"help",        no_argument,        0,  'h'},
        {"type",        optional_argument,  0,  't'},
        {"mode",        optional_argument,  0,  'm'},
        {"divisor",     optional_argument,  0,  'd'},
        {"range",       optional_argument,  0,  'r'},
        {"pin",         optional_argument,  0,  'p'},
        {"pulsewidth",  optional_argument,  0,  'w'},
         {0, 0, 0, 0}
    };
    /* getopt_long stores the option index here. */
    int option_index = 0;

    // Get command line options
    while (1)
    {
            c = getopt_long (argc, argv, "h::t::m::d::c::p::w::",
                             long_options, &option_index);
                         
            /* Detect the end of the options. */
            if (c == -1)
                break;
                               
            switch (c)
             {
             case 'h':
                 if (help_flag) {
                     printUsage();
                     exit(0);
                 }  
             case 't':
                 pwmType = optarg;
                 break;
             case 'm':
                pwmMode = optarg;
                break;
              case 'd':
                divisor = atoi(optarg);
                break;
              case 'r':
                range = atoi(optarg);
                break;
              case 'p':
                pin = atoi(optarg);
                globalPin = pin;
                break;
              case 'w':
                pulsewidth = atoi(optarg);
                break;
              case '?':
                 /* getopt_long already printed an error message. */
                 break;
              default:
                //printUsage();
                //exit(0);
                printf("Using default values\n");
              }
    }

    printf("Using: PWM pin: %d, PWM Type %s:, PWM Mode: %s, divisor: %d, range: %d, pulsewidth: %d\n", 
            pin, pwmType, pwmMode, divisor, range, pulsewidth);

    // Set up the WiringPi library, obtain resources, etc
    if (wiringPiSetup() == -1) {
        printf("setup wiringPi failed!");
        return 1;
    }

    if (strcmp(pwmType, hardware) == 0) {
        // register signal handler to catch ctl-C used to end the program for hardware PWM
        signal(SIGINT, hardwareInterruptHandler);
        runHardwarePWM(pwmMode, divisor, range, pin, pulsewidth);
    } else {
        // register interrupt handler to catch ctl-C used to end the program for software PWM
        signal(SIGINT, softwareInterruptHandler);
        runSoftwarePWM(pin, range, pulsewidth);
    }
}

// softwareInterruptHandler catches SIGINT when ctl-C is pressed in order to halt the program gracefully.
void softwareInterruptHandler(int sig) {
    // Had to switch to OUTPUT explicitly instead of using 'softPwmWrite()' as that wasn't reliable.
    pinMode(globalPin, OUTPUT);
    digitalWrite(globalPin, LOW);
    
    printf("\nExiting...\n");
    exit(0);
}

// hardwareInterruptHandler catches SIGINT when ctl-C is pressed in order to halt the program gracefully.
void hardwareInterruptHandler(int sig) {
    // Turn off LED
    pinMode(globalPin, PWM_OUTPUT);
    pwmWrite(globalPin, OFF);

    printf("\nExiting...\n");

    exit(0);
}

// runHardwarePWM starts PWM on a hardware PWM pin
void runHardwarePWM(char* pwmMode, int divisor, int range, int pin, int pulsewidth) {
    pinMode(pin, PWM_OUTPUT);
    pwmSetRange(range);
    if (strcmp(pwmMode, balanced) == 0) {
        pwmSetMode(PWM_MODE_BAL);
    } else {
        pwmSetMode(PWM_MODE_MS);
    }
    pwmSetClock(divisor);
    pwmWrite(pin, pulsewidth);
    while (1) {
        delayMicroseconds(1000);
    }
}

// runSoftwarePWM starts PWM implemented by this function vs. via the board hardware.
// NOTE: WiringPi directly implements software PWM. The PWM runs at 100MHz.
void runSoftwarePWM(int pin, int range, int pulsewidth) {
    softPwmCreate(pin, 0, range);
    softPwmWrite(pin, pulsewidth);
    while (1) {
        delayMicroseconds(1000);
    }
}
